PUBSTWO-1702 (instead of giving up with an error, just send a super-m…

orosz-usgs opened the pull request:
…inimal DOI record)

    New crossref template: body_minimal.ftlx
    New test resource: testOneNumberedSeriesPub-min.xml
    New test resource: testOneUnNumberedSeriesPub-min.xml
    Refcatored CrossrefTransformerIT.writeResylte to try two transformations, using full
template and minimal template
    Updated test: CrossrefTransformerIT

Commit:
PUBSTWO-1702 (instead of giving up with an error, just send a super-minimal DOI record)
    New crossref template: body_minimal.ftlx
    New test resource: testOneNumberedSeriesPub-min.xml
    New test resource: testOneUnNumberedSeriesPub-min.xml
    Refcatored CrossrefTransformerIT.writeResylte to try two transformations, using full
template and minimal template
    Updated test: CrossrefTransformerIT

dsteinich commented on src/main/java/gov/usgs/cida/pubs/transform/CrossrefTransformer.java:
```diff
@@ -130,49 +130,85 @@ public void write(Object result) throws IOException {
+		}
+
+		//add error message as a comment to the xml document
+		if(!xmlComment.isEmpty()) {
```
Do we really want the comment added every time a minimal record is sent? If so, would it be appropriate to add the comment to the body_minimal.ftlx version of the "report-paper"?

orosz-usgs replied:
Jim indicated it would be ok, I'll put the comment in body_minimal.ftlx

dsteinich commented on src/main/java/gov/usgs/cida/pubs/transform/CrossrefTransformer.java:
```diff
@@ -130,49 +130,85 @@ public void write(Object result) throws IOException {
 	 */
-	protected boolean writeResult(Publication<?> pub) throws IOException {
-		boolean success = false;
+	protected void writeResult(Publication<?> pub) throws IOException {
```
This seems overly complicated to me. Will getMinimalRecordComment need to be updated every time the values sent to the template get updated?

Perhaps just have two steps - try to write the expanded document and if that fails try to write the abbreviated document.

orosz-usgs replied:
The only time getMinimalRecordComment would need to be updated is to give
the user a specific message, otherwise it should be handled with a generic message.
will do the two steps.

dsteinich commented on src/main/java/gov/usgs/cida/pubs/transform/CrossrefTransformer.java:
```diff
@@ -130,49 +130,85 @@ public void write(Object result) throws IOException {
-			//add error message as a comment to the xml document
-			writeComment(message);
+	// return the reason the minimal record created or empty string if the minimal record does not need to be created.
+	protected String getMinimalRecordComment(Publication<?> pub, List<PublicationContributor<?>> contributors) {
```
The contributors are part of the pub, why are they being sent as a separate parameter?

orosz-usgs replied:
just over optimizing to avoid a second call to getContributors(pub)
That method probably belongs in the Publication class

dsteinich replied:
It is unclear to me why the original coder of this did not use the existing Publication.getContributorsToMap()

dsteinich commented on src/main/java/gov/usgs/cida/pubs/transform/CrossrefTransformer.java:
```diff
@@ -130,49 +130,85 @@ public void write(Object result) throws IOException {
+			comment = "Minimal doi record shown due to publication missing series title.";
+		} else if(isNumberedSeries && pub.getSeriesTitle().getOnlineIssn() == null) {
+			comment = "Minimal doi record shown due to publication missing series online issn number.";
 		}
```
What about the other potentially null values? Should they all be checked?

dsteinich commented on src/main/java/gov/usgs/cida/pubs/transform/CrossrefTransformer.java:
```diff
@@ -130,49 +130,85 @@ public void write(Object result) throws IOException {
+		return comment;
+	}
+
+	protected Map<String, Object> makeModel(Publication<?> pub, List<PublicationContributor<?>> contributors) {
```
I missed this in the pull request for PUBSTWO-1701 - the pubBusService.getWarehousePage(pub) (previously known as getIndexPage) should have never been in the business service. It is a property of the publication and the code should have resided there. Doing so would eliminate the need to inject the pubBusService into this class. I would not be adverse to this change happening in a seperate JIRA that could also look at creating properties on the publication that make use of the PubsUtils.is*() series of checks.

dsteinich replied:
As above (getMinimalRecordComment), contributors is part of the pubs object

dsteinich commented on src/main/java/gov/usgs/cida/pubs/transform/CrossrefTransformer.java:
```diff
@@ -188,7 +224,11 @@ protected void writeComment(String message) throws IOException {
 	}
-	
+
+	protected String getPubIdMsg(Publication<?> pub) {
```
I'm not a fan of all these one line methods, makes it harder to determine what is really going on.

Commit:
PUBSTWO-1702 (instead of giving up with an error, just send a super-minimal DOI record)
   Code cleanup:
     Removed authorKey and editorKey from CrossrefTransformer constructor
     Elimated methods getPubIdMsg and getExcludedErrorMessage from CrossrefTransformer
     Always try to create full record first, then minaml record if full failed.
     Moved creating minimal record comment to template
     Added missing check for getMinalRecComment and updated tests

dsteinich commented on src/main/java/gov/usgs/cida/pubs/transform/CrossrefTransformer.java:
```diff
@@ -130,49 +124,72 @@ public void write(Object result) throws IOException {
+		if(createMinimal) {
+			String logMessage = "";
+			try {
+				model.put("reasonForMinRec", getMinimalRecordComment(pub));
```
Isn't the reason for the minimal record really now the TemplateException? Yes, it could be the things checked in getMinimalRecordComment(), but it might not be.
I guess I wasn't clear before, in my opinion, getMinimalRecordComment() is trying to duplicate the functionality that the templating tool provides. Thus it is a form of duplicate/redundant code.

orosz-usgs replied:
Yes, getMinimalRecordComment() duplicates checks the templating tool performs. However, this makes it way easier for a support person or user to see that the full record failed because onlineIssn was missing by simply looking at the document. Otherwise, someone with access to the logs has to search for the error, and parse out field name from the freemarker stack trace.

orosz-usgs replied:
Ideally we would use template exception getMessageWithoutStackTop() for the xml comment, that is is probably not quite user friendly. I am ok with taking it out, since ideally the, minimal record should just be a rare exception.

dsteinich commented on src/main/java/gov/usgs/cida/pubs/transform/CrossrefTransformer.java:
```diff
@@ -130,49 +124,72 @@ public void write(Object result) throws IOException {
+
+		model.put("authorKey", ContributorType.AUTHORS);
+		model.put("editorKey", ContributorType.EDITORS);
+		model.put("compilerKey", ContributorType.COMPILERS);
```
probably should wait for the other JIRA to add this (and we actually have a lot of compilers in the system)

orosz-usgs replied:
yes, we have PUBSTWO-1721 ticket for add Compiler contributors. This was part of
the previous incomplete implementation of Compiler support. I'll take out compilerKey
of the model.

Commit:
PUBSTWO-1702 (instead of giving up with an error, just send a super-minimal DOI record)
   Code cleanup:
     Removed compilerKey from the freemarker model and template. This support
is slated for a future ticket.

Commit:
PUBSTWO-1702 (instead of giving up with an error, just send a super-minimal DOI record)
   Code cleanup:
     Removed comment from the minial Crossref xml.
